Test branch

ur10 opened the pull request:
These changes provide support for isolating thread-stacks. It is still a work in progress, assembly code for context restore and switch has some bugs.

ur10 committed:
BSP changes

- This provides the low-level support for protected thread-stack allocation

ur10 committed:
cpukit changes

- Protected thread stacks are allocated by configuring stack allocation in application to bsp_stack_allocate*
- prot_stack_* is used to set attributes of allocated memory regions
- memory_entries_* will have to be defined at the BSP specific level for flags translation, memory attributes setting/unsetting

ur10 committed:
Implemented the chain mechanism

Unsetting of memory entry at the lower level remains

ur10 committed:
cpukit changes for un-setting memory attributes

ur10 committed:
BSP changes for un-setting memory attributes

We simply assign the memory regions to be un-set as no-access

ur10 committed:
Small bug fix

ur10 committed:
Move stack management header to score

ur10 committed:
Support for context switch and moved stack management APIS to score

These functions should be called during context switch, initialization and restore

ur10 committed:
Low-level support for context-switch and moved memory management APIs to score

ur10 committed:
Bug fixes

ur10 committed:
Assembly suppport for context switch and bug fixes

ur10 committed:
Assembly suppport for context switch and bug fixes

> context switching and restoring support in arm/../cpu_asm.S
> threadloadenv.c loads protected stacks
> bug fixes in chain iteration in stackmanagement.c

heshamelmatary commented on bsps/arm/xilinx-zynq/mmu/bsp-set-mmu-attr.c:
```diff
@@ -0,0 +1,68 @@
+  break;
+  
+  default:
+    break;
```
you still need to return some integer value that the caller can check, of fatal error in here

heshamelmatary commented on bsps/arm/xilinx-zynq/mmu/bsp-set-mmu-attr.c:
```diff
@@ -0,0 +1,68 @@
+  break;
+
+  case READ_ONLY:
+  flags = ARMV7_MMU_READ_ONLY;
```
indentation

heshamelmatary commented on bsps/shared/start/stackalloc.c:
```diff
@@ -43,15 +44,19 @@ void bsp_stack_allocate_init(size_t stack_space_size)
 {
   void *stack = NULL;
-
+  uint32_t *page_table_base;
```
That's a shared code between all RTEMS ports. `uint32_t` won't be correct for 64-bit architecture. It's better to make it `uintptr_t`

heshamelmatary commented on bsps/shared/start/stackalloc.c:
```diff
@@ -43,15 +44,19 @@ void bsp_stack_allocate_init(size_t stack_space_size)
-
+  
+  page_table_base = 0x1000;   // This is hard-coded page table base address
+  prot_stack_allocate(stack, size, page_table_base);  // The current way to allocate protected stack is to assign memory attributes 
```
You may want to separate the calls to the allocated stack memory and setting the permissions on it, that's you call `prot_stack_allocate` from where `bsp_stack_allocate_init` is called. Otherwise, it will be confusing how to get the `page_table_base` from

ur10 replied:
I am a bit confused about this, `prot_stack_allocate` registers the attributes of the allocated stack to a chain, if we call it from `bsp_stack_allocate_init` how do we register the stack attributes structure to the chain? I agree we should definitely not pass the `page_table_base` from here, at least not in this format. I suppose I will have to think about as to how to decide what the `page_table_base` value for each stack should be.

heshamelmatary commented on cpukit/include/rtems/score/memorymanagement.h:
```diff
@@ -0,0 +1,25 @@
+    NO_ACCESS
+} memory_flags;
+
+void memory_entries_set(uint32_t *begin_addr, size_t size, memory_flags flags);
```
That's shared code. Assuming addresses are pointers to `uint32_t` isn't correct for 64-bit architectures. Please make it `uintptr_t`

heshamelmatary commented on cpukit/include/rtems/score/memorymanagement.h:
```diff
@@ -0,0 +1,25 @@
+
+void memory_entries_set(uint32_t *begin_addr, size_t size, memory_flags flags);
+
+void memory_entries_unset(uint32_t *begin_addr, size_t size);
```
ditto

heshamelmatary commented on cpukit/include/rtems/score/stackmanagement.h:
```diff
@@ -0,0 +1,60 @@
+typedef struct stack_attr
+{
+  Chain_Node    node; 
+  void*         stack_address;
```
for consistency, please make all addresses types `uintptr_t`

heshamelmatary commented on cpukit/include/rtems/score/stackmanagement.h:
```diff
@@ -0,0 +1,60 @@
+  Chain_Node    node; 
+  void*         stack_address;
+  size_t        size;
+  uint32_t      page_table_base;
```
`uintptr_t`

heshamelmatary commented on cpukit/score/cpu/arm/cpu.c:
```diff
@@ -31,6 +31,7 @@
 #include <rtems/score/thread.h>
 #include <rtems/score/tls.h>
 
+
```
avoid unnecessary white spaces

heshamelmatary commented on cpukit/score/cpu/arm/cpu_asm.S:
```diff
@@ -36,7 +36,7 @@
 
         .text
-
+		
```
whitespaces

heshamelmatary commented on cpukit/score/cpu/arm/cpu_asm.S:
```diff
@@ -65,7 +65,10 @@ DEFINE_FUNCTION_ARM(_CPU_Context_switch)
-
+ // #if defined ( USE_THREAD_STACK_PROTECTION )
+	str r0, [r0, #44]
+	bl  prot_stack_context_switch
```
How are you passing the argument to this function?

ur10 replied:
This is a mistake which I have fixed on my local branch, the instruction should be `ldr r0, [r0, #ARM_STACK_PROT_ATTR_OFFSET]`. This way the argument will be passed through the r0 register.

heshamelmatary replied:
But even after your fix, the code won’t behave correctly after the function returns as you’ve overwritten ‘r0’

ur10 replied:
Oh, sorry for the confusion. The complete code is
`mov r1, r0`
`ldr r0, [r0,  #ARM_STACK_PROT_ATTR_OFFSET]`
`bl prot_stack_context_switch`
`mov r0, r1`

heshamelmatary replied:
No problem, thanks for posting the code

heshamelmatary commented on cpukit/score/cpu/arm/cpu_asm.S:
```diff
@@ -65,7 +65,10 @@ DEFINE_FUNCTION_ARM(_CPU_Context_switch)
 	str	r3, [r0, #ARM_CONTEXT_CONTROL_ISR_DISPATCH_DISABLE]
-
+ // #if defined ( USE_THREAD_STACK_PROTECTION )
+	str r0, [r0, #44]
```
what's that supposed to be doing?

heshamelmatary replied:
Is see it's `ARM_STACK_PROT_ATTR_OFFSET`, please use the macro instead

heshamelmatary commented on cpukit/score/cpu/arm/include/rtems/score/cpu.h:
```diff
@@ -183,8 +184,11 @@
 
 #define ARM_EXCEPTION_FRAME_VFP_CONTEXT_OFFSET 72
 
+#define ARM_STACK_PROT_ATTR_OFFSET  44
```
Please make that `ARM_CONTEXT_STACK_PROT_ATTR_OFFSET` Also, there are a few #ifdefs that, if defined, will make this offset incorrect. You would want to to #ifdef it too

ur10 committed:
Isolate thread stacks

> Assembly bugs for context-switch fixed
> Implementation leaks memory when rtems user task is created, working for POSIX threads

ur10 committed:
Formatting fixes

ur10 committed:
Merge branch 'test_branch' of https://github.com/ur10/rtems into test_branch
